fix: added input validation in admin registration

MahendraDani opened the pull request:
<!--
This section can be deleted after reading.

We employ the following branching strategy to simplify the development process and to ensure that only stable code is pushed to the `master` branch:

- `develop`: For unstable code: New features and bug fixes.
- `master`: Where the stable production ready code lies. Only security related bugs.

NOTE!!!

ONLY SUBMIT PRS AGAINST OUR `DEVELOP` BRANCH. THE DEFAULT IS `MAIN`, SO YOU WILL HAVE TO MODIFY THIS BEFORE SUBMITTING YOUR PR FOR REVIEW. PRS MADE AGAINST `MAIN` WILL BE CLOSED.
-->

<!--
Thanks for submitting a pull request! Please provide enough information so that others can review your pull request.
-->

**What kind of change does this PR introduce?**
bug fix
<!-- E.g. a bugfix, feature, refactoring, etc… -->

**Issue Number:**
fixes #1222 

**Did you add tests for your changes?**
No
<!--Yes or No. Note: Add unit tests or automation tests for your code.-->

**Snapshots/Videos:**

<!--Add snapshots or videos wherever possible.-->



### When user registers with incorrect details

https://github.com/PalisadoesFoundation/talawa-admin/assets/116940083/6ac0a9a7-1242-4c4d-9770-f1aad8f5b0d5

- The user is not added in database and an toast with warning is given to the user.

### When user registers with valid details

https://github.com/PalisadoesFoundation/talawa-admin/assets/116940083/c1ed2d95-00d4-4578-b43a-38d329711260

- The user is successfully added in database if details are valid

**If relevant, did you update the documentation?**
No

<!--Add link to Talawa-Docs.-->

**Summary**

<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->
<!-- Try to link to an open issue for more information. -->
1. Installed [Zod](https://zod.dev/) library for input validation.
2. Added validations on `firstname`, `lastname`, `email` and `password` fields.
3. If validation is success the user is successfully registered.
4. If validation fail the user given a warning to fill the details correctly.

**Does this PR introduce a breaking change?**
No
<!-- If this PR introduces a breaking change, please describe the impact and a migration path for existing applications. -->

**Other information**

<!--Add extra information about this PR here-->

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-admin/blob/master/CONTRIBUTING.md)?**

<!--Yes or No-->
Yes

Kevoniat commented:
@MahendraDani  The Code Coverage  is down  please Investigate

MahendraDani commented:
@Kevoniat Added test coverage for the changes I made. Please let me know if there are any changes.
Thanks
<img width="811" alt="image" src="https://github.com/PalisadoesFoundation/talawa-admin/assets/116940083/9252e6f3-6bec-4387-8e74-0675f0c64303">

noman2002 commented on package.json:
```diff
@@ -66,7 +66,8 @@
     "web-vitals": "^1.0.1",
-    "yarn": "^1.22.17"
+    "yarn": "^1.22.17",
+    "zod": "^3.22.4"
```
Why this package is added ?

MahendraDani replied:
@noman2002 [Zod](https://zod.dev/) is used for validation of input given by the user.

noman2002 replied:
But we already have validation in various places without any package and its a very basic thing. I think it is not required to use any package for validation.

@aashimawadhwa your thoughts on this.

aashimawadhwa replied:
yes i think we can make validation work without the zod package too. @MahendraDani @noman2002

MahendraDani replied:
> yes i think we can make validation work without the zod package too. @MahendraDani @noman2002

@aashimawadhwa @noman2002 
I guess you are right. Should I close this PR?

MahendraDani replied:
@aashimawadhwa @noman2002 
My aim was to validate input for special characters and other characters like `[`, `]`, `{`, `}`,`@`, `#` etc. 
But it seems that when we validate any input by string `zod` considers the above characters in string too so using `zod` does not help that much.

rishav-jha-mech replied:
@MahendraDani why are you thinking on closing this PR ?, you can just remove zod and use TS or Regex to validate the strings

MahendraDani replied:
Will it be fine if I use regex and remove the package `zod` ?

rishav-jha-mech replied:
Yeah sure go ahead !

MahendraDani commented:
@palisadoes Would you please review this PR?

palisadoes commented:
There are reviewed already signed. Please work with them

aashimawadhwa commented:
@MahendraDani  please add proper validation messages the message provided right now is very generic. the validation should work in hierarchy that is the warning should be the first name cannot be number and so on .

MahendraDani commented:
> @MahendraDani please add proper validation messages the message provided right now is very generic. the validation should work in hierarchy that is the warning should be the first name cannot be number and so on .

I will remove `zod` package and use regex for validation and add more clear messages.
cc @aashimawadhwa @rishav-jha-mech

MahendraDani commented:
Currently we do not have any validation for `password`, should I also add the following validation for `password`:
1. Length should be more than 9
2. Should contain atleast one lowercase character: `a-z`
3. Should contain atleast one uppercase character : `A-Z`
4. Should contain atleast one special character : `!@#$^&*`
5. Should contain atleast one number :`0-9`

cc @aashimawadhwa @rishav-jha-mech

rishav-jha-mech commented:
Password length >= 6 is okay, if it's greater than 9, people will tend to forget their password, other validation steps are good to go

MahendraDani commented:
> Password length >= 6 is okay, if it's greater than 9, people will tend to forget their password, other validation steps are good to go

Alright, Thanks

MahendraDani commented:
### Changes introduced in https://github.com/PalisadoesFoundation/talawa-admin/pull/1247/commits/8b36c82c3ffc04a3c0d305eda1af6634c7b9d37d
- Removed `zod` package and used `Regular Expressions` for input validation
- Added input validation for `firstname`, `lastname`, `email`,`password`.
- Added more specific warning message for validation failures.

MahendraDani commented:
After resolving merge conflicts when I am trying to commit changes I am getting this following error:
```
/bin/sh: markdown-toc: command not found
node:internal/errors:932
  const err = new Error(message);
              ^

Error: Command failed: markdown-toc -i "CODE_OF_CONDUCT.md" --bullets "-"
    at checkExecSyncError (node:child_process:890:11)
    at execSync (node:child_process:962:15)
    at file:///Users/mahendradani/Documents/Programming/GSOC24/Palisadoes/talawa-admin/scripts/githooks/update-toc.js:10:3
    at Array.forEach (<anonymous>)
    at file:///Users/mahendradani/Documents/Programming/GSOC24/Palisadoes/talawa-admin/scripts/githooks/update-toc.js:8:15
    at ModuleJob.run (node:internal/modules/esm/module_job:218:25)
    at async ModuleLoader.import (node:internal/modules/esm/loader:329:24)
    at async loadESM (node:internal/process/esm_loader:34:7)
    at async handleMainPromise (node:internal/modules/run_main:113:12) {
  status: 127,
  signal: null,
  output: [ null, null, null ],
  pid: 18954,
  stdout: null,
  stderr: null
}

Node.js v20.10.0
husky - pre-commit hook exited with code 1 (error)
```

cc @aashimawadhwa @rishav-jha-mech

MahendraDani commented:
> After resolving merge conflicts when I am trying to commit changes I am getting this following error:
> 
> ```
> /bin/sh: markdown-toc: command not found
> node:internal/errors:932
>   const err = new Error(message);
>               ^
> 
> Error: Command failed: markdown-toc -i "CODE_OF_CONDUCT.md" --bullets "-"
>     at checkExecSyncError (node:child_process:890:11)
>     at execSync (node:child_process:962:15)
>     at file:///Users/mahendradani/Documents/Programming/GSOC24/Palisadoes/talawa-admin/scripts/githooks/update-toc.js:10:3
>     at Array.forEach (<anonymous>)
>     at file:///Users/mahendradani/Documents/Programming/GSOC24/Palisadoes/talawa-admin/scripts/githooks/update-toc.js:8:15
>     at ModuleJob.run (node:internal/modules/esm/module_job:218:25)
>     at async ModuleLoader.import (node:internal/modules/esm/loader:329:24)
>     at async loadESM (node:internal/process/esm_loader:34:7)
>     at async handleMainPromise (node:internal/modules/run_main:113:12) {
>   status: 127,
>   signal: null,
>   output: [ null, null, null ],
>   pid: 18954,
>   stdout: null,
>   stderr: null
> }
> 
> Node.js v20.10.0
> husky - pre-commit hook exited with code 1 (error)
> ```
> 
> cc @aashimawadhwa @rishav-jha-mech

I think the above error is because of the following command in `pacakage.json` file
```
"update:toc": "node scripts/githooks/update-toc.js"
```

rishav-jha-mech commented on package-lock.json:
```diff
@@ -25150,9 +25150,9 @@
-      "version": "3.21.4",
-      "resolved": "https://registry.npmjs.org/zod/-/zod-3.21.4.tgz",
-      "integrity": "sha512-m46AKbrzKVzOzs/DZgVnG5H55N1sv1M8qZU3A8RIKbs3mrACDNeIOeilDymVb2HdmP8uwshOCF4uJ8uM9rCqJw==",
+      "version": "3.22.4",
```
Revert change in this file

MahendraDani replied:
@rishav-jha-mech It is not working after reverting the files
I also tried to create a new branch and commit changes in that branch but it is also giving same error message

rishav-jha-mech replied:
@MahendraDani in any case if we are not using zod, there is no point in upgrading its version in `package-lock.json`, you are saying its not working after changing the files, what do you mean by that ?, also have you tried deleting the `node_modules` folder and reinstalling packages ?

MahendraDani replied:
@rishav-jha-mech I first unistalled `zod` using `npm uninstall --legacy-peer-deps zod` and then made the changes in `LoginPage.tsx` file.

rishav-jha-mech commented on public/locales/en.json:
```diff
@@ -9,7 +9,11 @@
-    "atleast_8_char_long": "Atleast 8 Character long",
+    "atleast_6_char_long": "Atleast 6 Character long",
+    "firstName_invalid": "First name should contain only lower and upper case letters",
+    "lastName_invalid": "Last name should contain only lower and upper case letters",
```
Add the same messages in other languages too

rishav-jha-mech commented:
@MahendraDani please remove the merge conflicts
